### PR TITLE
8284544: [Win] Name-Property of Spinner cannot be changed

### PR DESCRIPTION
**Pull request by arapte**
Currently we use the value of spinner as it's `UIA_NamePropertyId` when a11y client application requests for it.
Ideally we should use the text set by `Node.setAccessibleText()` as the `UIA_NamePropertyId`.
For other controls such as Slider, ListView we use the text set by setAccessibleText() API.

Fix:
Use the text set by `Node.setAccessibleText()` as the `UIA_NamePropertyId`.
This means, when a11y client requests `UIA_NamePropertyId`, which is mapped to AccessibleAttribute.TEXT attribute,  we shall return the accessible text.
So we need another way to read out the VALUE of the Spinner.
- For this we need to implement `IValueProvider` pattern for Spinner control
- IValueProvider requests the value of the control using it's API `get_ValueString()`
- It required to introduce a new AccessibleAttribute `VALUE_STRING`
- IValueProvider  also reads out if the control is editable or not, hence added `EDITABLE `case in `Spinner.queryAccessibleAttribute()`

Verification:
- Run any spinner app, with setAccessibleText set on spinner 
- Run Windows narrator and observe
- Without this fix:
- 1. Narrator does not read the text set by setAccessibleText
- 2. In application "Accessibility Insights for Windows", you can see the value of Spinner as the Name property
- After this fix:
- 1. Narrator reads the accessible text then value of Spinner and then if editable or not
- 2. In application "Accessibility Insights for Windows", you can see the text set by `setAccessibleText()` for Spinner as the Name property and the Spinner's value as value property

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change requires CSR request [JDK-8320505](https://bugs.openjdk.org/browse/JDK-8320505) to be approved
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issues
 * [JDK-8284544](https://bugs.openjdk.org/browse/JDK-8284544): [Win] Name-Property of Spinner cannot be changed (**Bug** - P3)
 * [JDK-8320505](https://bugs.openjdk.org/browse/JDK-8320505): [Win] Name-Property of Spinner cannot be changed (**CSR**)

### Reviewers
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Ajit Ghaisas](https://openjdk.org/census#aghaisas) (@aghaisas - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1291/head:pull/1291` \
`$ git checkout pull/1291`

Update a local copy of the PR: \
`$ git checkout pull/1291` \
`$ git pull https://git.openjdk.org/jfx.git pull/1291/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1291`

View PR using the GUI difftool: \
`$ git pr show -t 1291`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1291.diff">https://git.openjdk.org/jfx/pull/1291.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1291#issuecomment-1818264436)

### CODE REVIEW

**Comment by kevinrushforth** [comment]
This should be tested in connection with PR #1226

File: modules/javafx.controls/src/main/java/javafx/scene/control/Spinner.java [hunk 1]
```diff
@@ -845,6 +845,15 @@ public Object queryAccessibleAttribute(AccessibleAttribute attribute, Object...
                 }
                 return value != null ? value.toString() : "";
             }
+
+            case TEXT: {
+                String accText = getAccessibleText();
+                return (accText != null) ? accText : "";
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/main/java/javafx/scene/control/Spinner.java:851]: Is there a better default than the empty string? What do we do for Slider and other similar controls?

File: modules/javafx.graphics/src/main/java/javafx/scene/AccessibleAttribute.java [hunk 2]
```diff
@@ -774,6 +774,17 @@ public enum AccessibleAttribute {
...
     VALUE(Double.class),
 
+    /**
+     * Returns the value as string for the node.
+     * <ul>
+     * <li>Used by: Spinner </li>
+     * <li>Needs notify: yes </li>
+     * <li>Return Type: {@link String} </li>
+     * <li>Parameters: </li>
+     * </ul>
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/javafx/scene/AccessibleAttribute.java:784]: This will need `@since 22`

**Review by kevinrushforth — COMMENTED** [review]
I left quick comment and a question inline.

File: modules/javafx.controls/src/main/java/javafx/scene/control/Spinner.java (hunk 1 above)
Comment (arapte) [reply modules/javafx.controls/src/main/java/javafx/scene/control/Spinner.java:851]: With this fix, a11y client app reads a spinner as,
- when accessible text is set: [Accessible Text] Spinner [Value] [Editable / read only]
- when accessible text is not set: Spinner [Value] [Editable / read only]

Other controls like ComboBox, ChoiceBox return the value of control when accessible text is not set.
If we do the same for spinner, then the Spinner gets read as:  [Value] Spinner [Value] [Editable / read only]
In this case the value gets read twice.

The way spinner is read when accessible text is not set, it seems Ok to skip accessible text when it is not set.

**Comment by arapte** [comment]
Updated the PR with changes required for macOS, such that the Spinner is read same as before.

**Review by kevinrushforth — APPROVED** [review]
The changes look good and everything works as I would expect. I left one question, but it is fine as you have it.

File: modules/javafx.graphics/src/main/java/com/sun/glass/ui/mac/MacAccessible.java [hunk 3]
```diff
@@ -801,6 +801,11 @@ public void sendNotification(AccessibleAttribute notification) {
                     macNotification = MacNotification.NSAccessibilityValueChangedNotification;
                 }
                 break;
+            case VALUE_STRING:
+                if (getAttribute(ROLE) == AccessibleRole.SPINNER) {
```
[the remark this answers is not in the document]
Comment (arapte) [reply modules/javafx.graphics/src/main/java/com/sun/glass/ui/mac/MacAccessible.java:805]: Yes Kevin, as of now VALUE_STRING is only used for Spinner, but it's a pattern that is followed in this file: If anything is specific to a control we check the role. It also helps in finding code specific to a control. Let's please keep the check.

**Review by aghaisas — APPROVED** [review]
The fix looks good!

**Pull request closed without merging**